File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/java_to_cpp/code_output/TwentyFourPointGame.cpp
#include <vector>
#include <string>
#include <random>
#include <cctype>
#include <stdexcept>

class TwentyFourPointGame {
private:
    std::vector<int> nums;
    std::mt19937 rng{std::random_device{}()};

    std::vector<int> getMyCards() {
        std::vector<int> cards;
        std::uniform_int_distribution<int> dist(1, 9);
        for (int i = 0; i < 4; i++) {
            cards.push_back(dist(rng));
        }
        return cards;
    }

public:
    TwentyFourPointGame() {
        nums = getMyCards();
    }

    bool answer(const std::string& expression) {
        if (expression == "pass") {
            nums = getMyCards();
            return false;
        }

        int counts[10] = {};
        for (char c : expression) {
            if (std::isdigit(static_cast<unsigned char>(c))) {
                counts[c - '0']++;
            }
        }

        for (int num : nums) {
            if (counts[num] > 0) {
                counts[num]--;
            } else {
                return false;
            }
        }

        for (int count : counts) {
            if (count != 0) {
                return false;
            }
        }

        return evaluateExpression(expression);
    }

    bool evaluateExpression(const std::string& expression) {
        try {
            struct Parser {
                const std::string& expr;
                int pos = -1;
                int ch = 0;

                Parser(const std::string& e) : expr(e) {}

                void nextChar() {
                    ch = (++pos < static_cast<int>(expr.size())) ? expr[pos] : -1;
                }

                bool eat(int charToEat) {
                    while (ch == ' ') nextChar();
                    if (ch == charToEat) {
                        nextChar();
                        return true;
                    }
                    return false;
                }

                double parse() {
                    nextChar();
                    double x = parseExpression();
                    if (pos < static_cast<int>(expr.size())) throw std::runtime_error("Unexpected");
                    return x;
                }

                double parseExpression() {
                    double x = parseTerm();
                    for (;;) {
                        if (eat('+')) x += parseTerm();
                        else if (eat('-')) x -= parseTerm();
                        else return x;
                    }
                }

                double parseTerm() {
                    double x = parseFactor();
                    for (;;) {
                        if (eat('*')) x *= parseFactor();
                        else if (eat('/')) x /= parseFactor();
                        else return x;
                    }
                }

                double parseFactor() {
                    if (eat('+')) return parseFactor();
                    if (eat('-')) return -parseFactor();

                    double x;
                    int startPos = this->pos;
                    if (eat('(')) {
                        x = parseExpression();
                        eat(')');
                    } else if ((ch >= '0' && ch <= '9') || ch == '.') {
                        while ((ch >= '0' && ch <= '9') || ch == '.') nextChar();
                        x = std::stod(expr.substr(startPos, this->pos - startPos));
                    } else {
                        throw std::runtime_error("Unexpected");
                    }

                    return x;
                }
            };

            Parser parser(expression);
            return parser.parse() == 24;
        } catch (...) {
            return false;
        }
    }

    std::vector<int> getNums() {
        return nums;
    }
};

File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_py/code_output/TwentyFourPointGame.py
import random

class TwentyFourPointGame:
    def __init__(self):
        self.nums = self.get_my_cards()

    def get_my_cards(self):
        return [random.randint(1, 9) for _ in range(4)]

    def answer(self, expression):
        if expression == "pass":
            self.nums = self.get_my_cards()
            return False

        counts = [0] * 10
        for ch in expression:
            if '0' <= ch <= '9':
                counts[ord(ch) - ord('0')] += 1

        for num in self.nums:
            if counts[num] > 0:
                counts[num] -= 1
            else:
                return False

        if any(c != 0 for c in counts):
            return False

        return self.evaluate_expression(expression)

    def evaluate_expression(self, expression):
        class Parser:
            def __init__(self, s):
                self.s = s
                self.pos = -1
                self.ch = None
                self._next()

            def _next(self):
                self.pos += 1
                if self.pos < len(self.s):
                    self.ch = self.s[self.pos]
                else:
                    self.ch = None

            def eat(self, c):
                while self.ch is not None and self.ch == ' ':
                    self._next()
                if self.ch == c:
                    self._next()
                    return True
                return False

            def parse(self):
                self._next()
                x = self._parse_expression()
                if self.ch is not None:
                    raise Exception("Unexpected: " + str(self.ch))
                return x

            def _parse_expression(self):
                x = self._parse_term()
                while True:
                    if self.eat('+'):
                        x += self._parse_term()
                    elif self.eat('-'):
                        x -= self._parse_term()
                    else:
                        return x

            def _parse_term(self):
                x = self._parse_factor()
                while True:
                    if self.eat('*'):
                        x *= self._parse_factor()
                    elif self.eat('/'):
                        x /= self._parse_factor()
                    else:
                        return x

            def _parse_factor(self):
                if self.eat('+'):
                    return self._parse_factor()
                if self.eat('-'):
                    return -self._parse_factor()

                if self.eat('('):
                    x = self._parse_expression()
                    self.eat(')')
                    return x

                start_pos = self.pos
                if self.ch is not None and (('0' <= self.ch <= '9') or self.ch == '.'):
                    while self.ch is not None and (('0' <= self.ch <= '9') or self.ch == '.'):
                        self._next()
                    x = float(self.s[start_pos:self.pos])
                    return x
                else:
                    raise Exception("Unexpected: " + str(self.ch))

        try:
            return Parser(expression).parse() == 24.0
        except Exception:
            return False

    def get_nums(self):
        return self.nums